### ISSUE TITLE
dead_code lint false positive

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#45614]
```
fn main() {
    struct Foo<T> { x: T }
    type Bar = Foo<u32>;
    let spam = |Bar { x }| x != 0;
    println!("{}", spam(Foo { x: 10 }));
}
```

Gives:

```
warning: type alias is never used: `Bar`
 --> ...\test.rs:3:5
  |
3 |     type Bar = Foo<u32>;
  |     ^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(dead_code)] on by default
```

**Comment by sinkuu** [comment]
Minimized:

```rust
fn main() {
    struct Foo{} // warning: struct is never used
    let _ = |Foo{}| ();
}
```

So `dead_code` lint ignores type paths mentioned in patterns (2ec795b4f030ec19cf2ddc48fe5e2e158d5c0e42).

**Comment by scottmcm** [comment]
Repro without closure:
```rust
fn main() {
    struct Foo{} // warning: struct is never used
    let Foo{} = return;
}
```

**Comment by petrochenkov** [comment]
Code in https://github.com/rust-lang/rust/issues/45614#issuecomment-340261972 and https://github.com/rust-lang/rust/issues/45614#issuecomment-340600263 looks pretty dead. At least the structure can be considered dead if it is never constructed.

The original example in https://github.com/rust-lang/rust/issues/45614#issue-269387885 - the structure is constructed there and only liveness of the type alias `Bar` is misinterpreted, not of the structure itself.
Alias should be considered used whenever it's *named*, not necessarily constructed.
